src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java

```diff
@@ -191,6 +213,179 @@ public SearchResultPageDTO<QuizExercise> getAllOnPageWithSize(final PageableSear
         return new SearchResultPageDTO<>(exercisePage.getContent(), exercisePage.getTotalPages());
     }
 
+    /**
+     * Verifies that for DragAndDropQuestions all files are present and valid. Saves the files and updates the exercise accordingly.
+     *
+     * @param quizExercise the quiz exercise to create
+     * @param files        the provided files
+     */
+    public void handleDndQuizFileCreation(QuizExercise quizExercise, List<MultipartFile> files) throws IOException {
+        List<MultipartFile> nullsafeFiles = files == null ? new ArrayList<>() : files;
+        validateQuizExerciseFiles(quizExercise, nullsafeFiles, true);
+        Map<String, MultipartFile> fileMap = nullsafeFiles.stream().collect(Collectors.toMap(MultipartFile::getOriginalFilename, file -> file));
+
+        for (var question : quizExercise.getQuizQuestions()) {
+            if (question instanceof DragAndDropQuestion dragAndDropQuestion) {
+                if (dragAndDropQuestion.getBackgroundFilePath() != null) {
+                    saveDndQuestionBackground(dragAndDropQuestion, fileMap, null);
+                }
+                handleDndQuizDragItemsCreation(dragAndDropQuestion, fileMap);
+            }
+        }
+    }
+
+    private void handleDndQuizDragItemsCreation(DragAndDropQuestion dragAndDropQuestion, Map<String, MultipartFile> fileMap) throws IOException {
+        for (var dragItem : dragAndDropQuestion.getDragItems()) {
+            if (dragItem.getPictureFilePath() != null) {
+                saveDndDragItemPicture(dragItem, fileMap, null);
+            }
+        }
+    }
+
+    /**
+     * Verifies that for DragAndDropQuestions all files are present and valid. Saves the files and updates the exercise accordingly.
+     * Ignores unchanged paths and removes deleted background images.
+     *
+     * @param updatedExercise  the updated quiz exercise
+     * @param originalExercise the original quiz exercise
+     * @param files            the provided files
+     */
+    public void handleDndQuizFileUpdates(QuizExercise updatedExercise, QuizExercise originalExercise, List<MultipartFile> files) throws IOException {
+        List<MultipartFile> nullsafeFiles = files == null ? new ArrayList<>() : files;
+        validateQuizExerciseFiles(updatedExercise, nullsafeFiles, false);
+        // Find old drag items paths
+        Set<String> oldDragItemPaths = originalExercise.getQuizQuestions().stream().filter(question -> question instanceof DragAndDropQuestion)
+                .flatMap(question -> ((DragAndDropQuestion) question).getDragItems().stream()).map(DragItem::getPictureFilePath).filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+        Set<String> oldBackgroundPaths = originalExercise.getQuizQuestions().stream().filter(question -> question instanceof DragAndDropQuestion)
+                .map(question -> ((DragAndDropQuestion) question).getBackgroundFilePath()).filter(Objects::nonNull).collect(Collectors.toSet());
```

Comment:
How about first getting a set of all DnD questions?

Reply:
I could extract the functionality from this and another method and implement it another way. But good suggestion.